Use server side pagination for package list in project view

bgeuken opened the pull request:
We created a new endpoint for retrieving the package list, which
was needed for feeding it to datatable.

Co-authored-by: David Kang <dkang@suse.com>
Co-authored-by: Eduardo Navarro <enavarro@suse.com>



<!---
If you haven't done so already, please read the CONTRIBUTING.md file to learn how we work and what we expect from all contributors.

https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md
-->

bgeuken commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Server Side Pagination Everywhere](https://trello.com/c/ChgVZ4d6/1047-server-side-pagination-everywhere)

bgeuken commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Server Side Pagination Everywhere](https://trello.com/c/ChgVZ4d6/1047-server-side-pagination-everywhere)

coolo commented:
how does this work with inherited packages? And if you move things to javascript, shouldn't you add functional tests?

Ana06 commented on src/api/app/datatables/package_datatable.rb:
```diff
@@ -0,0 +1,38 @@
+    }
+  end
+
+  # rubocop:disable Naming/AccessorMethodName
```
I would exclude the datatables directory instead of doing it inline in all the files

bgeuken replied:
Disabling the whole file because we want to ignore a cop in one line, or a few, sounds like overkill to me.

Ana06 replied:
I mean to exclude `src/api/app/datatables` for `Naming/AccessorMethodName` in `.rubocop` (with a nice comment explaining why we are not going to finish this)

bgeuken replied:
Problem is that this would require me to convert all `Naming/AccessorMethodName` offenses we have right now and there are to many.

Ana06 replied:
I think this should be straightforward (or there is a bug in Rubocop). But this is not that important, do not block this PR for this. I will take a look into it and send a PR if I find a solution. Thanks for the branch.

Ana06 commented on src/api/app/datatables/package_datatable.rb:
```diff
@@ -0,0 +1,38 @@
+  def data
+    records.map do |record|
+      {
+        name: link_to(record.name, package_show_path(package: record, project: record.project)),
```
using `@project` instead `record.project` should save us many sql queries :thinking:

bgeuken replied:
Not really. This is taken care of by rails caching. The only query that happens here is loading the packages.

Ana06 replied:
Even if Rails catch the SQL query, still Rails need to check it is cached, while you have already this info stored in a variable available here.

bgeuken replied:
I am very reluctant to do micro-optimizations, but since I know that this isn't leading us anywhere I change it...

Ana06 commented on src/api/app/views/webui2/webui/project/_project_packages.html.haml:
```diff
@@ -1,18 +1,12 @@
 .card-body
```
This partial is called for both inherited packages (`@ipackages`) and packages. You can not remove things from here that easily. Also the variable `package` given as parameter when calling this partial is not used anymore.

bgeuken replied:
Oh, right. We forgot to split this up.

Ana06 replied:
you are right, there are two partials. I checked it wrong. Sorry.

But still you need to remove the `package` variable from:
` = render partial: 'project_packages', locals: { project: @project, packages: @packages }`

bgeuken replied:
Right now this partial still has a dependency on `@packages`. One could argue that we could, or even should, always show the package table. But I don't think we need to block this PR for that.

Ana06 replied:
ok

bgeuken commented:
@Ana06 When we triied this out we didn't remove the rubocop disable tags. So we didn't see the offense. I moved the test code to a [branch](https://github.com/bgeuken/open-build-service/tree/server_side_packages_rubocop) in case you want to play around with it and switched the code back to the initial version.
